core/src/main/java/org/elasticsearch/common/io/stream/Writeable.java

```diff
@@ -38,14 +43,20 @@
      */
     void writeTo(StreamOutput out) throws IOException;
 
-     @FunctionalInterface
-     interface IOFunction<T, R> {
-         /**
-          * Applies this function to the given argument.
-          *
-          * @param t the function argument
-          * @return the function result
-          */
-         R apply(T t) throws IOException;
-     }
+    @Override
+    default T readFrom(StreamInput in) throws IOException {
+        // See class javadoc for reasoning
+        throw new UnsupportedOperationException("Prefer calling a constructor that takes a StreamInput to calling readFrom.");
+    }
+
+    @FunctionalInterface
+    interface Reader<R> {
```

Comment:
I believe Function doesn't allow you to throw the IOException. I like it better this way because it is more specific - this is a function meant to read stuff.